File: app/src/main/java/space/lopatkin/spb/helpboardgamecard/di/ApplicationModule.java
package space.lopatkin.spb.helpboardgamecard.di;

import android.app.Application;
import dagger.Module;
import dagger.Provides;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.DeleteHelpcardUnlockedByHelpcardIdUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.DeleteHelpcardUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.DeleteHelpcardsByUnlockStateUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.GetAllHelpcardsUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.GetHelpcardByHelpcardIdUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.GetKeyboardTypeUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.SaveHelpcardNewByHelpcardIdUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.SaveKeyboardTypeByUserChoiceUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.UpdateHelpcardByHelpcardIdUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.UpdateHelpcardFavoriteByHelpcardIdUseCase;
import space.lopatkin.spb.helpboardgamecard.domain.usecase.UpdateHelpcardLockingByHelpcardIdUseCase;
import space.lopatkin.spb.helpboardgamecard.presentation.ViewModelFactory;

import javax.inject.Singleton;

@Module
public class ApplicationModule {

    private Application applicationContext;

    public ApplicationModule(Application applicationContext) {
        this.applicationContext = applicationContext;
    }

    @Singleton
    @Provides
    public Application provideApplication() {
        return applicationContext;
    }

    @Singleton
    @Provides
    public ViewModelFactory provideViewModelFactory(GetHelpcardByHelpcardIdUseCase getHelpcardByHelpcardIdUseCase,
                                                    GetAllHelpcardsUseCase getAllHelpcardsUseCase,
                                                    DeleteHelpcardUseCase deleteHelpcardUseCase,
                                                    DeleteHelpcardUnlockedByHelpcardIdUseCase deleteHelpcardUnlockedByHelpcardIdUseCase,
                                                    UpdateHelpcardByHelpcardIdUseCase updateHelpcardByHelpcardIdUseCase,
                                                    UpdateHelpcardFavoriteByHelpcardIdUseCase updateHelpcardFavoriteByHelpcardIdUseCase,
                                                    UpdateHelpcardLockingByHelpcardIdUseCase updateHelpcardLockingByHelpcardIdUseCase,
                                                    DeleteHelpcardsByUnlockStateUseCase deleteHelpcardsByUnlockStateUseCase,
                                                    SaveHelpcardNewByHelpcardIdUseCase saveHelpcardNewByHelpcardIdUseCase,
                                                    SaveKeyboardTypeByUserChoiceUseCase saveKeyboardTypeByUserChoiceUseCase,
                                                    GetKeyboardTypeUseCase getKeyboardTypeUseCase) {
        return new ViewModelFactory(getHelpcardByHelpcardIdUseCase,
                getAllHelpcardsUseCase,
                deleteHelpcardUseCase,
                deleteHelpcardUnlockedByHelpcardIdUseCase,
                updateHelpcardByHelpcardIdUseCase,
                updateHelpcardFavoriteByHelpcardIdUseCase,
                updateHelpcardLockingByHelpcardIdUseCase,
                deleteHelpcardsByUnlockStateUseCase,
                saveHelpcardNewByHelpcardIdUseCase,
                saveKeyboardTypeByUserChoiceUseCase,
                getKeyboardTypeUseCase
        );
    }

}

File: app/src/main/java/space/lopatkin/spb/helpboardgamecard/di/ApplicationModule.kt
package space.lopatkin.spb.helpboardgamecard.di

import android.content.Context
import dagger.Module
import dagger.Provides
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainCoroutineDispatcher
import space.lopatkin.spb.helpboardgamecard.domain.usecase.*
import space.lopatkin.spb.helpboardgamecard.presentation.ViewModelFactory
import javax.inject.Singleton

@Module
class ApplicationModule(
    private val context: Context,
    private val scope: CoroutineScope
) {

    @Singleton
    @Provides
    fun provideContext(): Context {
        return context
    }

    @Singleton
    @Provides
    fun provideScope(): CoroutineScope {
        return scope
    }

    @Singleton
    @Provides
    fun provideDispatchers(): CoroutineDispatchers {
        return CoroutineDispatchers()
    }

    @Singleton
    @Provides
    fun provideViewModelFactory(
        getHelpcardByBoardgameIdUseCase: GetHelpcardByBoardgameIdUseCase,
        getAllBoardgamesInfoUseCase: GetAllBoardgamesInfoUseCase,
        getBoardgameRawByBoardgameIdUseCase: GetBoardgameRawByBoardgameIdUseCase,
        deleteBoardgameUnlockedByBoardgameIdUseCase: DeleteBoardgameUnlockedByBoardgameIdUseCase,
        updateBoardgameByBoardgameIdUseCase: UpdateBoardgameByBoardgameIdUseCase,
        updateBoardgameFavoriteByBoardgameIdUseCase: UpdateBoardgameFavoriteByBoardgameIdUseCase,
        updateBoardgameLockingByBoardgameIdUseCase: UpdateBoardgameLockingByBoardgameIdUseCase,
        deleteBoardgamesByUnlockStateUseCase: DeleteBoardgamesByUnlockStateUseCase,
        saveBoardgameNewByBoardgameIdUseCase: SaveBoardgameNewByBoardgameIdUseCase,
        saveKeyboardTypeByUserChoiceUseCase: SaveKeyboardTypeByUserChoiceUseCase,
        getKeyboardTypeUseCase: GetKeyboardTypeUseCase,
        dispatchers: CoroutineDispatchers
    ): ViewModelFactory {
        return ViewModelFactory(
            getHelpcardByBoardgameIdUseCase = getHelpcardByBoardgameIdUseCase,
            getAllBoardgamesInfoUseCase = getAllBoardgamesInfoUseCase,
            getBoardgameRawByBoardgameIdUseCase = getBoardgameRawByBoardgameIdUseCase,
            deleteBoardgameUnlockedByBoardgameIdUseCase = deleteBoardgameUnlockedByBoardgameIdUseCase,
            updateBoardgameByBoardgameIdUseCase = updateBoardgameByBoardgameIdUseCase,
            updateBoardgameFavoriteByBoardgameIdUseCase = updateBoardgameFavoriteByBoardgameIdUseCase,
            updateBoardgameLockingByBoardgameIdUseCase = updateBoardgameLockingByBoardgameIdUseCase,
            deleteBoardgamesByUnlockStateUseCase = deleteBoardgamesByUnlockStateUseCase,
            saveBoardgameNewByBoardgameIdUseCase = saveBoardgameNewByBoardgameIdUseCase,
            saveKeyboardTypeByUserChoiceUseCase = saveKeyboardTypeByUserChoiceUseCase,
            getKeyboardTypeUseCase = getKeyboardTypeUseCase,
            dispatchers = dispatchers
        )
    }

    class CoroutineDispatchers(
        val main: MainCoroutineDispatcher = Dispatchers.Main,
        val default: CoroutineDispatcher = Dispatchers.Default,
        val io: CoroutineDispatcher = Dispatchers.IO,
        val unconfined: CoroutineDispatcher = Dispatchers.Unconfined,
    )

}
